### PR TITLE
Update handling pending approve() transaction when performing swap #6193

### LINKED ISSUE #6193: During swap and if the user is prompted for ERC20 approve(), after transaction goes through, it should automatically prompt for the swap transaction
Maybe related to: #6055

1. Swap WETH -> MATIC (and assuming user has not approved or currently approved amount is not enough)
2. Tap Continue
3. Actionsheet for approve() appears
4. After Confirm for the approve() transaction and the transaction is confirmed https://polygonscan.com/tx/0xcf1070ce99203be4062a1d80569952ef3f05a653fc75e68c90fc1d3b9c92b165

Expected
---
5. The actionsheet for the swap transaction appears
6. Tap Confirm to send swap transaction

Observed
---
5. It stays at the swap screen (app still responsive, see screenshot), and doesn't move on to prompt for the swap transaction

6. Tapping the Continue button again displays the actionsheet for the swap transaction
7. Tap Confirm to send swap transaction (and the swap transactions works)

### PR DESCRIPTION
**Pull request by oa-s**
Closes #6193

### CODE REVIEW

**Comment by oa-s** [comment]
@hboon pr is ready for review

File: modules/AlphaWalletFoundation/AlphaWalletFoundation/EtherClient/Models/GetErc20Allowance.swift [hunk 1]
```diff
@@ -30,6 +30,7 @@ class GetErc20Allowance {
                 } else {
                     return (false, amount - allowance)
                 }
-            }.eraseToAnyPublisher()
+            }.receive(on: RunLoop.main)
```
Comment (hboon) [inline modules/AlphaWalletFoundation/AlphaWalletFoundation/EtherClient/Models/GetErc20Allowance.swift:33]: Why is the explict `.receive(on: .main)` needed?

File: modules/AlphaWalletFoundation/AlphaWalletFoundation/SwapToken/NativeSwap/Helpers/ApproveSwapProvider.swift [hunk 2]
```diff
@@ -83,28 +91,15 @@ public final class ApproveSwapProvider {
...
                     .map { _ in return true }
-                    .catch { error -> AnyPublisher<Bool, SwapError> in
-                        if error.embedded is WaitTillTransactionCompleted.NotCompletedYetError {
-                            return .fail(SwapError.approveTransactionNotCompleted)
-                        } else if let error = error.embedded as? SwapError {
-                            switch error {
-                            case .userCancelledApproval:
-                                return .just(false)
-                            case .unableToBuildSwapUnsignedTransactionFromSwapProvider, .invalidJson, .unableToBuildSwapUnsignedTransaction, .approveTransactionNotCompleted, .unknownError, .tokenOrSwapQuoteNotFound, .inner:
-                                return .fail(error)
```
Comment (hboon) [inline modules/AlphaWalletFoundation/AlphaWalletFoundation/SwapToken/NativeSwap/Helpers/ApproveSwapProvider.swift:104]: Removing this means we no longer display the error, right? Did we display the error previously?

**Review by hboon — COMMENTED** [review]
IMO, better to generally extract out refactorings and skip-review them (if you confident) and to isolate the fix in separate PRs. Easier to review and less likely to make mistakes.

File: modules/AlphaWalletFoundation/AlphaWalletFoundation/EtherClient/Models/GetErc20Allowance.swift (hunk 1 above)
Comment (oa-s) [reply modules/AlphaWalletFoundation/AlphaWalletFoundation/EtherClient/Models/GetErc20Allowance.swift:33]: by default it was on `.global`, updating the ui should be on `.main`

File: modules/AlphaWalletFoundation/AlphaWalletFoundation/SwapToken/NativeSwap/Helpers/ApproveSwapProvider.swift (hunk 2 above)
Comment (oa-s) [reply modules/AlphaWalletFoundation/AlphaWalletFoundation/SwapToken/NativeSwap/Helpers/ApproveSwapProvider.swift:104]: nope, that is just for waiting pending transaction, we didn't display error for it even before, the error that could be generated throws on top level, so we still able to display it

File: modules/AlphaWalletFoundation/AlphaWalletFoundation/EtherClient/Models/GetErc20Allowance.swift (hunk 1 above)
Comment (hboon) [reply modules/AlphaWalletFoundation/AlphaWalletFoundation/EtherClient/Models/GetErc20Allowance.swift:33]: But `callSmartContract(withServer:contract:functionName:abiString:parameters:shouldDelayIfCached:)` is on `main` right?

**Review by hboon — APPROVED** [review]

**Review by hboon — APPROVED** [review]

**Pull request merged**